src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -274,19 +274,41 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
             {
                 Requires.NotNull(operation, nameof(operation));
 
+                // Note: this code is duplicated and not extracted to minimize allocating extra async state machines.
+                // For performance reasons, we want to minimize the number of Joins performed, and also keep the size of the
+                // JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a JTF.RunAsync. This requires
+                // us to not ConfigureAwait(true) on the semaphore. However, that breaks our contract of resuming on the sync context
+                // you started on. To partially fix this, we will at least resume you on the main thread or thread pool.
```

Comment:
We don't really have a contract yet, as this is a new feature. So rather than say it breaks a contract, let's just state what our contract is. And for that matter, it would be good to express this in the xml doc comments (on the base class) for this method.